x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/ssl/SSLReloadDuringStartupIntegTests.java

```diff
@@ -59,66 +77,123 @@ protected boolean transportSSLEnabled() {
             @Override
             public Settings onNodeStopped(String nodeName) throws Exception {
-                Path origKeystorePath = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode.jks");
-                Path keystorePath = env.configFile().resolve("testnode.jks");
-                Path updatedKeystorePath = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode_updated.jks");
-                assertTrue(Files.exists(keystorePath));
-                // replace the keystore with one that will fail during handshaking
-                copyAndAtomicMoveIfPossible(updatedKeystorePath, keystorePath);
-                new Thread(() -> {
-                    latch.countDown();
-                    try {
-                        latch.await();
-                        Thread.sleep(randomLongBetween(1L, 2000L));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                    }
-
-                    // replace the bad keystore with a new copy of the good one
+                LOGGER.info("Node [{}] stopped in {}ms.", nodeName, (System.nanoTime() - stopNanos) / 1000000F);
+                atomicCopyIfPossible(badKeystorePath, nodeKeystorePath);
+                final Thread fixKeystoreThread = new Thread(() -> {
+                    waitUntilNodeStartupIsReadyToBegin(beforeKeystoreFix); // SYNCHRONIZED TASK
                     try {
-                        copyAndAtomicMoveIfPossible(origKeystorePath, keystorePath);
-                        writtenLatch.countDown();
+                        atomicCopyIfPossible(goodKeystorePath, nodeKeystorePath);
+                        afterKeystoreFix.countDown(); // Indicate keystore is fixed
                     } catch (IOException e) {
                         throw new UncheckedIOException(e);
                     }
-                }).start();
-
-                latch.countDown();
-                try {
-                    latch.await();
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                }
-                return super.onNodeStopped(nodeName);
+                });
+                fixKeystoreThread.setName("Fix Keystore");
+                fixKeystoreThread.start();
+                waitUntilFixKeystoreIsReadyToBegin(beforeKeystoreFix); // SYNCHRONIZED TASK
+                return super.onNodeStopped(nodeName); // Return control to RestartCallback to do startup next
             }
         });
-        writtenLatch.await();
-        ensureClusterSizeConsistency();
-        ensureFullyConnectedCluster();
+        final long awaitNanos = System.nanoTime(); // actual await nanos
+        try {
+            afterKeystoreFix.await(); // Wait until keystore is fixed
+        } finally {
+            LOGGER.info("Awaited {}ms. Verifying the cluster...", (System.nanoTime() - awaitNanos) / 1000000F);
```

Comment:
Are all of these timing necessary ?  It adds alot of visual clutter to the tests and it isn't clear to why the "took" timings here are relevant.

Reply:
Yes, the timings are necessary. The intermittent failures are timing issues. If one of these steps takes an expected long time, it might be a contributing factor.

Reply:
np .. however is it possible to clean up the timings. For example:
```class Scratch {
    public static void main(String[] args) {
        timed(method1(), "stringparam", "action1");
        timed(method2(), 10, "action2");
    }

    static <T> void timed(Consumer<T> consumer, T param, String action){
        long start = System.nanoTime();
        consumer.accept(param);
        System.out.println(action + " took " + (System.nanoTime() - start) + "ns");
    }
    static Consumer<String> method1() {
        return s -> System.out.println(s);
    }
    static Consumer<Integer> method2() {
        return i -> System.out.println(i);
    }
}
```

Reply:
I tried something slightly different, try-with-resource using a `Timed implements Closeable` instance. Let me know if that looks OK, or if you want me to refactor to something closer to your example above.

Reply:
I still have a preference towards wrapping these timings via a functional approach instead of using try/close since I think it reads cleaner. For example, here you need 3 lines vs. 1 line to add each timing, and then for each try with checked exception you need to handle the same exception in multiple places. IMO it makes it harder to read, but I also understand that it probably just a style choice, but I do think a functional approach is more inline with other code in ES.

Reply:
Done.